pkg/pool-stable/contracts/ComposableStablePoolProtocolFees.sol

```diff
@@ -227,44 +232,68 @@ abstract contract ComposableStablePoolProtocolFees is
+        // join or exit, while this code pays any protocol fees due on the current join or exit.
+        // The amp and rates are constant during a single transaction, so it doesn't matter if there
+        // is an ongoing amp change, and we can ignore yield.
+
+        // This represents the total value exchanged in this particular join-exit, including swap fees.
+        //
+        // Note that the "growth" is actually negative in the case of exits: both the supply and invariant
+        // decrease. Since swap fees are symmetrical, it is numerically equivalent (and conceptually simpler)
+        // to just invert the sign in the exit case, and use the same "growth" terminology and calculations
+        // for both.
+        //
+        uint256 invariantGrowth = (
+            postJoinExitInvariant > preJoinExitInvariant
+                ? postJoinExitInvariant - preJoinExitInvariant
+                : preJoinExitInvariant - postJoinExitInvariant
+        )
+            .divDown(preJoinExitInvariant);
+
+        // The BPT increase represents the value exchanged for the user's tokens in the join or exit.
+        // The more unbalanced the join or exit, the greater the "taxable" amount subject to swap fees,
+        // and the lower the BPT amount in or out.
+        //
+        // In a proportional join or exit, the taxable amount will be 0, so the BPT growth would
+        // equal the total growth. Otherwise, the BPT would be minted based on fewer tokens, and
+        // the BPT growth will be less than the total.
+        //
+        // As we do above with the invariant, since joins and exits are symmetrical, "growth" means the
+        // amount of increase (joins) or decrease (exits).
+        uint256 bptGrowth = (
+            postJoinExitSupply > preJoinExitSupply
+                ? postJoinExitSupply - preJoinExitSupply
+                : preJoinExitSupply - postJoinExitSupply
+        )
+            .divDown(preJoinExitSupply);
+
+        // Therefore, the difference between the invariant growth and bpt increase rates represents the amount
+        // that was charged swap fees, on which protocol fees are also due.
+
+        if (invariantGrowth > bptGrowth) {
+            // Since the growth values are already normalized, multiply by the protocol swap fee to convert
+            // to a percentage owned to the protocol.
+            uint256 protocolSwapFeePercentage = (invariantGrowth - bptGrowth).mulDown(
+                getProtocolFeePercentageCache(ProtocolFeeType.SWAP)
+            );
+
+            // Since this will be minted as BPT, which increases the total supply, we need to mint
+            // slightly more so that it reflects this percentage of the total supply after minting.
+            uint256 protocolFeeAmount = preJoinExitSupply.mulDown(protocolSwapFeePercentage).divDown(
+                protocolSwapFeePercentage.complement()
+            );
+
+            _payProtocolFees(protocolFeeAmount);
```

Comment:
I think this is wrong. Here's how I think about this (assuming join):

a) the supply increased by some ratio `SR`. the invariant can increase by `SR` without fees
b) `after invariant - pre invariant * SR` is the invariant delta due to fees, `fee delta inv`
c) to convert that to a pool percentage, we need to divide `fee delta inv` not by `pre invariant`, but by `after invariant`. the protocol swap percentage of this then is the protocol ownership %
d) we apply `pct = m / (s + m)`

Reply:
Looking into this now. I think we should also be using `postJoinExitSupply` on L283